### PR TITLE
Feature/role manager

### PR DESCRIPTION
**Pull request by Ayoush**
**Available endpoints** :

1. Added api which will reveal all the endpoint paths, action and feature so that while providing access user can select among those.
2. Added an api to display all those available endpoints

**APIs to add user group**: 

1. User group is a collection of policies which a user who is part of it will be able to access only those features in an application
2. APIs for creating a group, showing multiple groups, updating a group and deleting it is exposed

**Invitation flow altered**:

1. Invitation flow is altered and now while inviting admin can mention the upcoming user group or policies which will restrict the access provided to him/her.

**Policy addition on application boot up**:

1. Logic to fill policies automatically when some other developer exposes new api has been added which will compare and fetch out the new recent added policies and will fill it at the start of the application.

**Gateway Data Dump desc order**:

1. Now the data will come in descending order based on the inserted timestamp.

### CODE REVIEW

**Review by arjun289 — CHANGES_REQUESTED** [review]

File: apps/acqdat_api/lib/acqdat_api/api_access/user_group.ex [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  defdelegate get_all(data, preloads), to: UserGroup
+  defdelegate get(id), to: UserGroup
+  defdelegate delete(user_group), to: UserGroup
+
+  def create(params) do
+    params = params_extraction(params)
+    policy_ids = Policy.extract_policies(params.actions)
+    params = Map.put_new(params, :policy_ids, policy_ids)
+    params = Map.put_new(params, :user_ids, [])
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/api_access/user_group.ex:17]: done

File: apps/acqdat_api/lib/acqdat_api/api_access/user_group.ex [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    verify_group(UserGroup.create(params))
+  end
+
+  def update(group, params) do
+    params = for {key, val} <- params, into: %{}, do: {String.to_atom(key), val}
+    policy_ids = Policy.extract_policies(params.actions)
+    present_policy_ids = UserGroup.policies(group.id)
+    policy_ids = Enum.uniq(policy_ids ++ present_policy_ids)
+    params = Map.put_new(params, :policy_ids, policy_ids)
+    params = Map.put_new(params, :user_ids, [])
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/api_access/user_group.ex:27]: it is there as i kept it for future use as suggested by you. I am doing this just for ensuring other wise this check is also there in validators also.

File: apps/acqdat_api/lib/acqdat_api/entity-management/organisation.ex [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  defdelegate get(id), to: OrgModel
+  defdelegate update(org, data), to: OrgModel
+  defdelegate get_all(data, preloads), to: OrgModel
+  defdelegate delete(org), to: OrgModel
+
+  def create(params) do
+    params = params_extraction(params)
+
+    case params.user_details do
+      nil -> verify_organisation(OrgModel.create(params))
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/entity-management/organisation.ex:19]: Done, i made it compulsory in validator of organisation so this code will not be reached without providing user details

File: apps/acqdat_api/lib/acqdat_api/extract_routes.ex [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+  #   "DashboardManagement" => %{
+  #     "DashboardExport" => {
+  #       "controller" => "DashboardExportController",
+  #       "actions" => {
+  #           "export" => %{path: :export, request_type: :get},
+  #           "show" => %{path: :show, request_type: :get}
+  #       }
+  #     }
+  #  }
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/extract_routes.ex:15]: Done

File: apps/acqdat_api/lib/acqdat_api/extract_routes.ex [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+  def start_link(_opts) do
+    Agent.start_link(fn -> extract_routes() end, name: __MODULE__)
+  end
+
+  def value() do
+    Agent.get(__MODULE__, & &1)
+  end
+
+  def extract_routes() do
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/extract_routes.ex:25]: I am using this function in ExtractedRoutesController

File: apps/acqdat_api/lib/acqdat_api_web/auth_pipeline/api_access_auth.ex [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            acc ++ [policies.policy]
+          end) ++
+            Enum.reduce(user_policies, [], fn policies, acc ->
+              acc ++ [policies.policy]
+            end)
+
+        user_actions =
+          Enum.reduce(policies, [], fn policy, acc ->
+            acc ++ [policy |> Map.from_struct() |> Map.drop([:_id, :__meta__])]
+          end)
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/auth_pipeline/api_access_auth.ex]: done

File: apps/acqdat_api/lib/acqdat_api_web/auth_pipeline/api_access_auth.ex [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    case is_nil(List.first(group_ids)) do
+      false ->
+        group_policies = extract_policies(group_ids)
+
+        policies =
+          Enum.reduce(group_policies, [], fn policies, acc ->
+            acc ++ [policies.policy]
+          end) ++
+            Enum.reduce(user_policies, [], fn policies, acc ->
+              acc ++ [policies.policy]
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/auth_pipeline/api_access_auth.ex]: Done

**Review by arjun289 — APPROVED** [review]

**Pull request merged**